lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -52,500 +28,823 @@
+    if (this_chunk_sz > EP0_BUF_SZ) {
+        this_chunk_sz = EP0_BUF_SZ;
+    }
 
+    // 4-byte-at-a-time copy
+    unsigned int this_chunk_sz_words = this_chunk_sz / 4;
+    for (unsigned int i = 0; i < this_chunk_sz_words; i++) {
+        HWREG(USB0_BASE + USB_O_FIFO0) = setup_data_to_send[i];
+    }
+    setup_data_to_send += this_chunk_sz_words;
+
+    // Copy remainder
+    for (unsigned int i = 0; i < this_chunk_sz % 4; i++) {
+        HWREGB(USB0_BASE + USB_O_FIFO0) = ((const uint8_t *)setup_data_to_send)[i];
+    }
+
+    setup_data_to_send_sz -= this_chunk_sz;
 }
 
-// *****************************************************************************
-//
-// Handles CDC driver notifications related to the receive channel (data from
-// the USB host).
-//
-// \param ulCBData is the client-supplied callback data value for this channel.
-// \param ulEvent identifies the event we are being notified about.
-// \param ulMsgValue is an event-specific value.
-// \param pvMsgData is an event-specific pointer.
-//
-// This function is called by the CDC driver to notify us of any events
-// related to operation of the receive data channel (the OUT channel carrying
-// data from the USB host).
-//
-// \return The return value is event-specific.
-//
-// *****************************************************************************
-unsigned int RxHandler(void *pvCBData, unsigned int ulEvent, unsigned int ulMsgValue, void *pvMsgData) {
+static void usb_device_intr() {
+    IntSystemStatusClear(SYS_INT_USB0);
+    uint32_t intr_src = HWREG(USB_0_OTGBASE + USB_0_INTR_SRC);
 
-    //
-    // Which event are we being sent?
-    //
-    switch (ulEvent)
-    {
-        //
-        // A new packet has been received.
-        //
-        case USB_EVENT_RX_AVAILABLE: {
-            // Hack to get CTRL+C to work. Needs to be moved to pbsys.
-            char *test = pvMsgData;
-            extern bool pbsys_main_stdin_event(uint8_t c);
-            for (unsigned int i = 0; i < ulMsgValue; i++) {
-                pbsys_main_stdin_event(test[i]);
-            }
-            break;
+    if (intr_src & (1 << 18)) {
+        // USB reset
+
+        // Reset state variables
+        usb_addr = 0;
+        usb_config = 0;
+        setup_data_to_send = 0;
+        usb_addr_needs_setting = 0;
+
+        if (HWREGH(USB0_BASE + USB_O_POWER) & (1 << 4)) {
+            is_usb_hs = 1;
+        } else {
+            is_usb_hs = 0;
         }
 
-        //
-        // We are being asked how much unprocessed data we have still to
-        // process. We return 0 if the UART is currently idle or 1 if it is
-        // in the process of transmitting something. The actual number of
-        // bytes in the UART FIFO is not important here, merely whether or
-        // not everything previously sent to us has been transmitted.
-        //
-        case USB_EVENT_DATA_REMAINING: {
-            return 0;
+        // Set up the FIFOs
```

Comment:
This seems like a lot to do in an interrupt handler. Is it something we can defer? i.e. set a flag and poll the protothread/process to do it.

Reply:
This is *visually* a lot of code, but if you actually look at what it does, it just does a series of register pokes. It's not really meaningfully longer than many of the other interrupt code paths.